Issue #1756: Moving QUANTILE

hawkfish opened the pull request:
* Check that the previous frame had no NULLs. 
* Add test for variable window discrete.
* DeValue the segment tree API

Commit:
Issue #1756: Moving QUANTILE

Check that the previous frame had no NULLs.

Commit:
Issue #1756: Moving QUANTILE

Add test for variable window discrete.

Commit:
Issue #1756: Moving QUANTILE

DeValue the segment tree API.

Mytherin commented on src/execution/operator/aggregate/physical_window.cpp:
```diff
@@ -663,8 +663,12 @@ static void ComputeWindowExpression(BoundWindowExpression *wexpr, ChunkCollectio
+			auto &target_chunk = output.GetChunkForRow(row_idx);
+			auto &target = target_chunk.data[output_col];
+			const auto target_offset = row_idx % STANDARD_VECTOR_SIZE;
+			VectorOperations::Copy(source, target, 1, 0, target_offset);
```
Perhaps we could let the `aggregate.finalize` take in a `result_offset` parameter. That way we can avoid the Copy entirely, as finalize can write the result directly into the vector at the correct position. That would also avoid the constant vector problems you mentioned.

hawkfish replied:
We can do that, but every `Finalize` method in the code will have to change.

Mytherin replied:
I think that should actually not be too difficult `AggregateExecutor::Finalize`:

```cpp
        // AggregateExecutor::Finalize
	template <class STATE_TYPE, class RESULT_TYPE, class OP>
	static void Finalize(Vector &states, FunctionData *bind_data, Vector &result, idx_t count) {
		if (states.GetVectorType() == VectorType::CONSTANT_VECTOR) {
			result.SetVectorType(VectorType::CONSTANT_VECTOR);

			auto sdata = ConstantVector::GetData<STATE_TYPE *>(states);
			auto rdata = ConstantVector::GetData<RESULT_TYPE>(result);
			OP::template Finalize<RESULT_TYPE, STATE_TYPE>(result, bind_data, *sdata, rdata,
			                                               ConstantVector::Validity(result), 0);
		} else {
			D_ASSERT(states.GetVectorType() == VectorType::FLAT_VECTOR);
			result.SetVectorType(VectorType::FLAT_VECTOR);

			auto sdata = FlatVector::GetData<STATE_TYPE *>(states);
			auto rdata = FlatVector::GetData<RESULT_TYPE>(result);
			for (idx_t i = 0; i < count; i++) {
				OP::template Finalize<RESULT_TYPE, STATE_TYPE>(result, bind_data, sdata[i], rdata,
				                                               FlatVector::Validity(result), i);
			}
		}
	}

```

And internally use a method such as this:

```cpp
	template <class T, class STATE>
	static void Finalize(Vector &result, FunctionData *, STATE *state, T *target, ValidityMask &mask, idx_t idx) {
		if (!state->isset) {
			mask.SetInvalid(idx);
		} else {
			target[idx] = state->value;
		}
	}
```

So we would only need to change AggregateExecutor::Finalize together with the (small number of) aggregates that manually implement the Finalize method. 

Anyway, it could also be done in a subsequent PR if you prefer.

Mytherin commented on src/function/aggregate/nested/histogram.cpp:
```diff
@@ -101,23 +101,30 @@ static void HistogramCombineFunction(Vector &state, Vector &combined, idx_t coun
 }
 
+static inline ValidityMask &AnyValidity(Vector &v) {
+	return (v.GetVectorType() == VectorType::CONSTANT_VECTOR) ? ConstantVector::Validity(v) : FlatVector::Validity(v);
```
This seems like it might cause problems since constant and flat vectors are semantically different. `FlatVector::Validity` used to do this switch, but we changed it since it hid certain types of bugs. `Finalize` should generally be called with the `result` set to a flat vector as that is the expected usage of the API (and function APIs in general).

Mytherin reviewed (commented):
Thanks for the PR! Looks great. Always nice to see some value removal. Some comments:

Commit:
Merge branch 'master' into hawkfish-moving-quantile

Commit:
Issue #1756: Moving QUANTILE

Write moving aggregates directly into the window results collection.

One part of this fix required extending the aggregate function APIs
to take an output offset (usually 0). This enables the aggregate API
to be used to build aggregate results incrementally.

A second more complicated change was to move the result chunk collection
construction from the Sink stage to the Scan stage. This was needed to
get rid of the unnatural situation where the aggregates were defaulting
to NULL, contrary to the contract. It also had the happy result that
these blocks of NULLs are not being reordered when the input gets
hashed and sorted.

hawkfish commented:
Not sure what happened, but it passed on other platforms and on my fork.

Mytherin commented:
The test failures are not related to this PR indeed. Thanks for the changes, looks great now!
